src/freenet/node/DNSRequester.java

```diff
@@ -56,23 +56,24 @@ public void run() {
     private void realRun() {
         PeerNode[] nodes = node.getPeers().myPeers();
         long now = System.currentTimeMillis();
-        if((now - lastLogTime) > 1000) {
-        	if(logMINOR)
+        if((now - lastLogTime) > 100) {
+        	if(logMINOR) {
         		Logger.minor(this, "Processing DNS Requests (log rate-limited)");
+            }
             lastLogTime = now;
         }
-        for(PeerNode pn: nodes) {
-            //Logger.minor(this, "Node: "+pn);
-            if(!pn.isConnected()) {
-                // Not connected
-                // Try new DNS lookup
-            	//Logger.minor(this, "Doing lookup on "+pn+" of "+nodes.length);
-                pn.maybeUpdateHandshakeIPs(false);
-            }
+        // check a randomly chosen node to avoid sending bursts of DNS requests
+        PeerNode pn = nodes[node.getFastWeakRandom().nextInt(nodes.length)];
```

Comment:
This approach turns the peer node DNS resolving into the [Coupon collector's problem](https://en.wikipedia.org/wiki/Coupon_collector%27s_problem). For 80 peer nodes, the expected number of tries before all are hit is 397 (taking roughly 8 minutes), and it takes 563 attempts (~10 minutes) before 95% probability is reached - even assuming the peer node list is static, which it is not. This may not be desirable.

This can be resolved by ignoring connected nodes prior to random selection.

Reply:
That’s a great catch — thank you!

Does only checking non-connected nodes really resolve it? I’d think it would only reduce the problem.

Reply:
@bertm the new commit should fix this: it uses a set and a deque to ensure that we only check a node again if at least 80% of the other unconnected nodes have been checked (as cheapskate’s ordering trick ☺).

It uses the location to check for equality, because these must not be equal, and because a set and a deque of doubles is cheap.

Reply:
See https://github.com/hyphanet/fred/pull/993/commits/ced3b161f8480ff570927be43450cd38e3443617